Multipart form data

ivarne opened the pull request:
Work in process

Builds on https://github.com/Altinn/app-lib-dotnet/pull/328, so temporary the target branch is different

## Related Issue(s)
- https://github.com/Altinn/app-lib-dotnet/issues/187
- closes #128

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

ivarne committed:
Let ModelDeserializer return a Result type instead of nullable

ivarne committed:
Remove NullDataProcessor and let user registrer multiple processors

ivarne committed:
Update IDataProcessor interface and support multipart data write

ivarne committed:
Update v7tov8 script for new IDataProcessor interface

Also ignore CS1998 in apps (warning when not awaiting anything in async hooks)

tjololo commented on src/Altinn.App.Api/Controllers/StatelessDataController.cs:
```diff
@@ -115,8 +115,12 @@ public class StatelessDataController : ControllerBase
+            {
+                _logger.LogInformation("ProcessDataRead for {modelType} using {dataProcesor}", appModel.GetType().Name, dataProcessor.GetType().Name);
+                await dataProcessor.ProcessDataRead(virtualInstance, null, appModel);
+            }
```
Could be extracted to a private method as the same lines of code are repeated multiple times.

ivarne replied:
Probably not a bad idea. I rejected it because to do it properly I'd need a new service to extract the common parts between stateless and normal.

```c#
internal class DataProcessorWrapper(IEnumerable<IDataProcessor> dataProcessors, ILogger<DataProcessorWrapper> logger)
{
    internal async Task DataProcessRead(...)
    {
        foreach(...)
        {
            // log
            // processRead(...)
        }
    }
    // ProcessDataWrite
}
```

But I did not really like the pattern.

tjololo commented on src/Altinn.App.Core/Features/DataProcessing/NullDataProcessor.cs:
👍

tjololo commented on src/Altinn.App.Core/Helpers/Serialization/ModelDeserializer.cs:
```diff
@@ -61,50 +58,74 @@ public ModelDeserializer(ILogger logger, Type modelType)
+            {
+                if (secondSection?.ContentDisposition?.Contains("name=\"previousValues\"") != true)
+                {
+                    return ModelDeserializerResult.FromError("First entry in mulipart serialization must have name=\"dataModel\"");
```
Missleading error returned
```suggestion
                    return ModelDeserializerResult.FromError("Second entry in mulipart serialization must have name=\"previousValues\"");
```

tjololo commented on src/Altinn.App.Core/Helpers/Serialization/ModelDeserializer.cs:
```diff
@@ -61,50 +58,74 @@ public ModelDeserializer(ILogger logger, Type modelType)
+                reportedChanges = await System.Text.Json.JsonSerializer.DeserializeAsync<Dictionary<string, string?>>(secondSection.Body);
+                if (await reader.ReadNextSectionAsync() != null)
+                {
+                    return ModelDeserializerResult.FromError("Multipart request had more than 2 elements");
```
Should the error be even more specific and say that we expect "dataModel" and optionally a "previousValues" as we know that we only support these?

ivarne replied:
Not sure if that is required. To get here we need the first element to be `dataModel` and the second to be `previousValues`, otherwise the earlier errors will trigger.

tjololo replied:
thats true!

tjololo commented on src/Altinn.App.Core/Helpers/Serialization/ModelDeserializer.cs:
Should we take the effort to remove Newtonsoft from this part of the code now that we have to test this part of the code anyways?
Our goal is still to migrate all code from newtonsoft to system.text

ivarne replied:
I think that should be its own issue and PR. It has consequences and we might need it to be a setting in case someone has issues.

ivarne replied:
Should probably look into it with v8

tjololo replied:
:+1: created a issue for this specific case: https://github.com/Altinn/app-lib-dotnet/issues/345

tjololo commented on src/Altinn.App.Core/Helpers/Serialization/ModelDeserializerResult.cs:
```diff
@@ -0,0 +1,64 @@
+    /// }
+    /// </code>
+    /// </example>
+    [MemberNotNullWhen(true, nameof(Error))]
+    [MemberNotNullWhen(false, nameof(Model))]
```
Nice! Did not know about this attribute, this is why I love reviews :)

ivarne replied:
It's quite new but really great for readability of the resulting code.

tjololo commented on src/Altinn.App.Core/Infrastructure/Clients/Storage/DataClient.cs:
```diff
@@ -171,15 +171,15 @@ public async Task<object> GetFormData(Guid instanceGuid, Type type, string org,
+                ModelDeserializerResult deserializerResult = await deserializer.DeserializeAsync(stream, "application/xml");
 
-                if (deserializer.Error != null || model is null)
+                if (deserializerResult.Error is not null || deserializerResult.Model is null )
```
Could HasError be used instead? That should cover the same checks if I read the code correctly

ivarne replied:
It probably should. I'll test and update.

tjololo reviewed (commented):
Nice work! :godmode:

ivarne committed:
Code review updates

ivarne reviewed (commented):
Fikset noen av kommentarene

tjololo commented on src/Altinn.App.Core/Helpers/Serialization/ModelDeserializer.cs:
```diff
@@ -61,50 +58,74 @@ public ModelDeserializer(ILogger logger, Type modelType)
-            }
+                _logger.LogError(ex, "Unexpected exception when attempting to deserialize JSON into '{modelType}'", _modelType);
+                return ModelDeserializerResult.FromError($"Unexpected exception when attempting to deserialize JSON into '{_modelType}'");
+            };
```
```suggestion
            }
```

tjololo reviewed (approved):
One last thing I missed, but looks good. Lets get this from review to test 🙁

ivarne committed:
Update src/Altinn.App.Core/Helpers/Serialization/ModelDeserializer.cs

Co-authored-by: Vemund Gaukstad <tjololo@users.noreply.github.com>

ivarne committed:
First draft of tests for controllers

ivarne committed:
Finish writing tests

ivarne commented:
@tjololo 😡 The tests works on OSX, but not in linux? Any idea how I might debug that issue?

It seems likely to be an issue with unreadable files due to case-sensitivity causing authentication failure, but I did not find anything.

tjololo commented:
I discovered the same thing yesterday on a branch I have. It's the same issue we have in app-template. I have a linux laptop so plan to spend the day investigating what the issue is

ivarne committed:
Set partyId in token to null

ivarne committed:
Add partyId correctly in test tokens

Also add a Dockerfile for running tests on linux

tjololo committed:
Fix bad case for roles folder in testsetup

tjololo commented:
Can we add some test for the bad requests. Like that the first part not having name=dataModel and so on?

olemartinorg commented:
I tried this against frontend v4/main now, and it's not quite there. Should be the same with v3, as I haven't tested with the form data rewrite in frontend yet.

1. A feature flag, `multiPartSave` is expected to be returned in applicationmetadata for frontend to start sending multipart changes. I added this manually in `FrontendFeatues.cs` when testing.
2. Frontend sends `name="dataModel"`, not `name=dataModel` (note the quotes). I manually added some quotes in `ModelDeserializer` when testing.
3. The whole data model seems to be returned, not the metadata combined with `changedFields`, as is expected in v3. Be sure not to conflate this task with #316, as that's an entirely different beast. Also, returning the entire data model on the top level might not be what we want, as it would be nice to include validation messages etc as well in the future.

ivarne committed:
Merge remote-tracking branch 'origin/v8' into ivarne/multipart-form-data

ivarne committed:
Use proper parsing library to decode multipart/form-data requests

ivarne committed:
Continue returning changed only changed values from data put.

ivarne committed:
Fix so that it compiles (still failing tests)

ivarne committed:
Fix tests by disabeling redirects in applciationfactory client

For some reason backend returns 303 See Other when datamodel updates are availible.
This is a redirect code and crashed the tests as no Location header was set.

ivarne committed:
Add more tests

ivarne committed:
More tests

olemartinorg commented:
Tested again now, and it's all working against v4/main as far as I can tell! :raised_hands:

tjololo commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -58,7 +58,7 @@ public class DataController : ControllerBase
         /// <param name="instantiationProcessor">Instantiation processor</param>
         /// <param name="dataClient">A service with access to data storage.</param>
-        /// <param name="dataProcessor">Serive implemnting logic during data read/write</param>
+        /// <param name="dataProcessors">Serive implemnting logic during data read/write</param>
```
```suggestion
        /// <param name="dataProcessors">Services implementing logic during data read/write</param>
```

tjololo reviewed (approved):
Have a look at the code smells/codeql warnings, and try to add the last few percentages of testscoverage and we should be good to go as far as I can see

ivarne committed:
Update src/Altinn.App.Api/Controllers/DataController.cs

Co-authored-by: Vemund Gaukstad <tjololo@users.noreply.github.com>

ivarne committed:
Fix code smells

ivarne committed:
Fix tests

ivarne committed:
Tests OK now?

ivarne committed:
Merge remote-tracking branch 'origin/v8' into ivarne/multipart-form-data

ivarne committed:
More tests for model deserializer

ivarne commented on src/Altinn.App.Api/Controllers/StatelessDataController.cs:
```diff
@@ -115,12 +115,24 @@ public class StatelessDataController : ControllerBase
-            Instance virutalInstance = new Instance() { InstanceOwner = owner };
-            await _dataProcessor.ProcessDataRead(virutalInstance, null, appModel);
+            Instance virtualInstance = new Instance() { InstanceOwner = owner };
+            await ProcessAllDataWrite(virtualInstance, appModel);
```
@tjololo This was not good!!!

Happens to be OK, as `ProcessAllDataWrite` actually calls `processDataRead` 😊 ☺️ I'll fix in the next rewrite that uses json patch
